File: typescript/protein-translation/protein-translation.ts
const codonToProtein: { [key: string]: string } = {
  AUG: 'Methionine',
  UUU: 'Phenylalanine',
  UUC: 'Phenylalanine',
  UUA: 'Leucine',
  UUG: 'Leucine',
  UCU: 'Serine',
  UCC: 'Serine',
  UCA: 'Serine',
  UCG: 'Serine',
  UAU: 'Tyrosine',
  UAC: 'Tyrosine',
  UGU: 'Cysteine',
  UGC: 'Cysteine',
  UGG: 'Tryptophan',
  UAA: 'STOP',
  UAG: 'STOP',
  UGA: 'STOP',
}

export function translate(rna: string): string[] {
  let proteins: string[] = []
  let rnaCopy: string = rna

  while(rnaCopy.length > 0) {
    let codon = rnaCopy.slice(0, 3)
    let protien = codonToProtein[codon]

    if(!protien) {
      throw new Error('Invalid codon')
    } else if(protien === 'STOP') {
      break
    } else {
      proteins = [...proteins, protien]
      rnaCopy = rnaCopy.slice(3)
    }
  }

  return proteins
}

File: ruby/protein-translation/protein_translation.rb
class InvalidCodonError < StandardError
end

module Translation
  # takes a strand of RNA and returns the proteins that are translated from the RNA
  # @param [String] strand
  # @return [Array<String>]
  def self.of_rna(strand)
    codons = strand.scan(/\w{3}|\w+/)
    protiens = []
    codons.each do |codon|
      case codon
      when "AUG"
        protiens << "Methionine"
      when "UUU", "UUC"
        protiens << "Phenylalanine"
      when "UUA", "UUG"
        protiens << "Leucine"
      when "UCU", "UCC", "UCA", "UCG"
        protiens << "Serine"
      when "UAU", "UAC"
        protiens << "Tyrosine"
      when "UGU", "UGC"
        protiens << "Cysteine"
      when "UGG"
        protiens << "Tryptophan"
      when "UAA", "UAG", "UGA"
        break
      else
        raise InvalidCodonError
      end
    end
    protiens
  end
end
